contents/huffman_encoding/code/kotlin/huffman.kt

```diff
@@ -0,0 +1,130 @@
+import java.util.*
+
+// node type
+sealed class Node(val frequency: Int): Comparable<Node> {
+  // we can sort nodes by frequency
+  override fun compareTo(other: Node) = frequency - other.frequency
+}
+class Leaf(freq: Int, val letter: Char): Node(freq)
+class Branch(freq: Int, val left: Node, val right: Node): Node(freq)
+
+class HuffmanTree (val textSample: String) {
+  val root: Node
+  val codeBook: CodeBook
+  
+  init {
+    // calculate frequencies
+    val frequencyMap = HashMap<Char, Int>()
+    for (char in textSample) {
+      val newFrequency = (frequencyMap.get(char) ?: 0) + 1
+      frequencyMap.put(char, newFrequency)
+    }
+    // populate the queue with leaves
+    val priorityQueue = PriorityQueue<Node>()
+    for (m in frequencyMap.entries) {
+      priorityQueue.add(Leaf(m.value, m.key))
+    }
+    // join leaves
+    while (priorityQueue.size > 1) {
+      val left = priorityQueue.remove()
+      val right = priorityQueue.remove()
+      priorityQueue.add(Branch(left.frequency + right.frequency, left, right))
+    }
+    // the root node is the last remaining leaf
+    root = priorityQueue.remove()
+    // initialize the code book
+    codeBook = CodeBook(root)
+  }
+  
+  // encode a string using the tree
+  fun encode(source: String): String {
```

Comment:
```suggestion
  fun encode(source: String) = source.map { codeBook.encode(it) }.joinToString(separator = "")
```
replaces the whole function.